Airflow tutorial to use functional DAGs

vikramkoka opened the pull request:
Created a new Airflow ETL tutorial to use functional DAGs. Also created a DAG to perform the same operations without using functional DAGs to be compatible with Airflow 1.10.x and to show the difference between the "functional" and the "classic" ways side by side to illustrate the differences. 

I tried to make this simple and not rely on any operators. I will create the html documentation page to cover this as soon as I get confirmation from @casassg and @turbaszek that this is in the right general direction. Also tagging @ashb for general review. 

closes: #9041 

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

vikramkoka committed:
Airflow tutorial to use Decorated flows

Created a new Airflow tutorial to use Decorated Flows (a.k.a. functional
DAGs). Also created a DAG to perform the same operations without using
functional DAGs to be compatible with Airflow 1.10.x and to show the
difference.

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+as part of the documentation that goes along with the Airflow Functional DAG tutorial located
+[here](https://airflow.apache.org/tutorial_functional_dag.html)
+"""
+# [START tutorial]
```
Will we use the whole example in docs? If yes this is ok 👌

vikramkoka replied:
Yes, I was planning on using the whole example in the docs, unless there were objections to that.

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+    'email_on_retry': False,
+    'retries': 1,
+    'retry_delay': timedelta(minutes=5),
+}
```
Should we limit this to minimum?

vikramkoka replied:
I am fine with that.

vikramkoka replied:
I just did a commit with these changes for both files.

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+    default_args=default_args,
+    description='ETL DAG tutorial',
+    schedule_interval=timedelta(days=1),
+    start_date=days_ago(2),
```
I'm personally in favor of using `None` as schedule in most of examples because when users turn on one the example DAG scheduler will not try to create multiple runs

vikramkoka replied:
That sentiment makes sense to me, I am fine with that. Will make the change

vikramkoka replied:
I just did a commit with these changes for both files.

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+# [END instantiate_dag]
+# [START documentation]
+    dag.doc_md = __doc__
+# [END documentation]
```
Should we move this to DAG invocation?

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+# [END load_function]
+
+# [START extract_task]
+    extract_task = PythonOperator(
+        task_id='extract',
+        python_callable=extract,
+    )
```
We should use `@task()` decorator instead of `PythonOperator`, that was the main point of functional DAGs

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+        python_callable=extract,
+    )
+# [END basic_task]
+    extract_task.doc_md = """\
```
If I'm not mistaken we should be able to do:
```
@task(doc_md="here goes the docs")
def extract(**kwargs):
    ...
```

turbaszek commented on airflow/example_dags/tutorial_etl_dag.py:
```diff
@@ -0,0 +1,134 @@
+# [END documentation]
+
+# [START extract_function]
+    def extract(**kwargs):
+        ti = kwargs['ti']
```
If we want to show the functional approach in full swing we should use:
```python
from airflow.operators.python import get_current_context

def extract():
    ctx = get_current_context()
    ti = ctx['ti']
    ...
```

turbaszek replied:
Ok, I see it 😄

turbaszek commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+        data_string = u'{"1001": 301.27, "1002": 433.21, "1003": 502.22}'
+        return data_string
+# [END extract]
+    extract.doc_md = """\
```
If I'm not mistaken we should be able to do:
```
@task(doc_md="here goes the docs")
def extract(**kwargs):
    ...
```

turbaszek commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+    """
+
+# [START load]
+    @dag.task()
```
I'm wondering if it would be better to define the task before DAG so they can be reusable? No strong opinion here

vikramkoka replied:
I would prefer to leave it inside the DAG, if that's ok with you.

turbaszek commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+    #### Transform task
+    A simple Transform task which takes in the collection of order data and computes
+    the total order value.
+    """
```
And there will be indentation unless we do:
```suggestion
#### Transform task
A simple Transform task which takes in the collection of order data and computes
the total order value.
"""
```

vikramkoka replied:
I just did a commit with these changes for all tasks in both files.

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+
+        total_value = {"total_order_value": total_order_value}
+        total_value_json_string = json.dumps(total_value)
+        return total_value_json_string
```
I would return directly the dictionary and use `multiple_outputs=True` instead. See: https://github.com/apache/airflow/blob/master/tests/operators/test_python.py#L357

vikramkoka replied:
That makes sense. I will make the change.

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+    """
+
+# [START transform]
+    @dag.task()
```
```suggestion
    @dag.task(multiple_outputs=True)
```

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+        for value in order_data.values():
+            total_order_value += value
+
+        total_value = {"total_order_value": total_order_value}
```
```suggestion
        return {"total_order_value": total_order_value}
```

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,119 @@
+
+    order_data = extract()
+    order_summary = transform(order_data)
+    load(order_summary)
```
```suggestion
    load(order_summary["total_order_value"])
```

vikramkoka committed:
Apply suggestions from code review

It makes sense to simplify the return variables being passed around without needlessly converting to JSON and then reconverting back.

Co-authored-by: Gerard Casas Saez <casassg@users.noreply.github.com>

vikramkoka committed:
Update tutorial_functional_etl_dag.py

Fixed data passing between tasks to be more natural without converting to JSON and converting back to variables.

vikramkoka committed:
Updated dag options and task doc formating

Based on feedback on the PR, updated the DAG options (including schedule) and the fixed the task documentation to avoid indentation.

vikramkoka committed:
Added documentation file for functional dag tutorial

Added the tutorial documentation to the docs directory. Fixed linting errors in the example dags.
Tweaked some doc references in the example dags for inclusion into the tutorial documentation.
Added the example dags to example tests.

ashb commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,110 @@
+# The DAG object; we'll need this to instantiate a DAG
+from airflow import DAG
+from airflow.utils.dates import days_ago
+from airflow.decorators import task
```
```suggestion
```

Since we are using `@dag.task`

ashb commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,110 @@
+    tags=['example'],
+) as dag:
+# [END instantiate_dag]
+# [START documentation]
```
These two lines are the source of the `expected an indented block (comment)` error -- I don't know if we can indent them without messing up the inclusion in docs. Hmmm

kaxil committed:
Fix Static checks

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,111 @@
+    # [END documentation]
+
+    # [START extract]
+    @dag.task(multiple_outputs=True)
```
This one doesn't need multiple_outputs

vikramkoka replied:
Thanks @casassg , got a little confused on this when reading the code in PythonOperator - changed this now.

vikramkoka committed:
Removed multiple_outputs from task defn

Had a multiple_outputs=True defined in the Extract task defn, which was unnecessary. - Removed based on feedback.

vikramkoka committed:
Fixed spell check issues

Added word summarization to the spelling list in the docs directory to
fix spell check error in tutorial. Also fixed another sentence error in
the tutorial document.

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,111 @@
+    # [END documentation]
+
+    # [START extract]
+    @dag.task()
```
You can do 
```suggestion
    @dag.task
```
which is cleaner imo

turbaszek replied:
I think this creates mypy issues https://github.com/apache/airflow/pull/10930#discussion_r489375827 :<

casassg commented on airflow/example_dags/tutorial_functional_etl_dag.py:
```diff
@@ -0,0 +1,111 @@
+"""
+
+    # [START load]
+    @dag.task()
```
```suggestion
    @dag.task
```

casassg commented on docs/tutorial_functional.rst:
```diff
@@ -0,0 +1,151 @@
+
+What's Next?
+-------------
+You have seen how simple it is to write DAGs using the Functional DAG paradigm within
```
Would be nice to have this point to https://github.com/apache/airflow/blob/master/docs/concepts.rst#functional-dags and https://github.com/apache/airflow/blob/master/docs/concepts.rst#functional-dags more specifically. 

Also, it may be interesting to link the AIP-31 for more insight into this and potentially my talk and Jonathan's article as learn more resources? CC @turbaszek for thoughts

turbaszek replied:
+1 for linking AIP and this article for more insight:
https://medium.com/databand-ai/aip-31-airflow-functional-dag-definition-b34852a632d0

vikramkoka replied:
I added the links to the AIP and to the Functional DAGs section of the Concept doc. 

I was not sure if it was ok to add a link to an external site which has a paywall (Medium) from the Apache open source docs, so did not add that.

ashb committed:
fixup! Airflow tutorial to use Decorated flows

vikramkoka reviewed (commented):
Looks good to me @ashb
